homcc/server/cache.py

```diff
@@ -4,31 +4,71 @@
 
 """Caching module of the homcc server."""
 import logging
+from collections import OrderedDict
 from pathlib import Path
 from threading import Lock
-from typing import Dict
 
 logger = logging.getLogger(__name__)
 
 
 class Cache:
     """Represents the homcc server cache that is used to cache dependencies."""
 
-    cache: Dict[str, str]
+    cache: OrderedDict[str, str]
     """'Hash' -> 'File path' on server map for holding paths to cached files"""
     cache_mutex: Lock
     """Mutex for locking the cache."""
     cache_folder: Path
     """Path to the cache on the file system."""
+    max_size_bytes: int
+    """Maximum size of the cache in bytes."""
+    current_size_bytes: int
+    """Current size of the cache in bytes."""
+
+    def __init__(self, root_folder: Path, max_size_bytes: int):
+        if max_size_bytes <= 0:
+            raise RuntimeError("Maximum size of cache must be strictly positive.")
 
-    def __init__(self, root_folder: Path):
         self.cache_folder = self._create_cache_folder(root_folder)
-        self.cache: Dict[str, str] = {}
+        self.cache: OrderedDict[str, str] = OrderedDict()
         self.cache_mutex: Lock = Lock()
+        self.max_size_bytes = max_size_bytes
+        self.current_size_bytes = 0
+
+    def _get_cache_file_path(self, hash_value: str) -> Path:
+        return self.cache_folder / hash_value
+
+    def __contains__(self, key: str):
+        with self.cache_mutex:
+            contained: bool = key in self.cache
+            if contained:
+                self.cache.move_to_end(key)
+
+            return contained
 
-    def __contains__(self, key):
+    def __len__(self) -> int:
         with self.cache_mutex:
-            return key in self.cache
+            return len(self.cache)
+
+    def _evict_oldest(self):
+        """
+        Evicts the oldest entry from the cache.
+        Note: The caller of this method has to ensure that the cache is locked.
+        """
+        oldest_hash, oldest_path_str = self.cache.popitem(last=False)
+        oldest_path = Path(oldest_path_str)
+        oldest_size = oldest_path.stat().st_size
```

Comment:
If the file does not exist, isn't this going to raise an exception?

Reply:
Very good point! This also affects cache size calculation. I moved the `stat` call into the try block, and adjusted the log message that the cache size calculation may be wrong if files are deleted by other processes.
A different solution would be to store the size of each cache entry in memory, but I don't really like that. In the end, some cache entries being deleted by other processes are probably rather very much an edge case, so I guess handling it like this is fine.